Controller/Adminhtml/Menu/Edit.php

```diff
@@ -1,58 +1,80 @@
 {
     const REGISTRY_CODE = 'snowmenu_menu';
+
     /**
-     * @var MenuRepositoryInterface
+     * @var PageFactory
      */
-    private $menuRepository;
+    private $resultPageFactory;
+
     /**
      * @var Registry
      */
-    private $registry;
+    private $coreRegistry;
 
-    public function __construct(Action\Context $context, MenuRepositoryInterface $menuRepository, Registry $registry)
-    {
-        parent::__construct($context);
-        $this->menuRepository = $menuRepository;
-        $this->registry = $registry;
+    /**
+     * @param Context $context
+     * @param Registry $coreRegistry
+     * @param PageFactory $resultPageFactory
+     * @param MenuRepositoryInterface $menuRepository
+     * @param MenuFactory $menuFactory
+     */
+    public function __construct(
+        Context $context,
+        Registry $coreRegistry,
+        PageFactory $resultPageFactory,
+        MenuRepositoryInterface $menuRepository,
+        MenuFactory $menuFactory
+    ) {
+        $this->resultPageFactory = $resultPageFactory;
+        $this->coreRegistry = $coreRegistry;
+        parent::__construct($context, $menuRepository, $menuFactory);
     }
 
     /**
-     * Dispatch request
-     *
-     * @return \Magento\Framework\Controller\ResultInterface|ResponseInterface
-     * @throws \Magento\Framework\Exception\NotFoundException
+     * @return ResultInterface
      */
-    public function execute()
+    public function execute(): ResultInterface
     {
-        $id = $this->getRequest()->getParam('id');
-        try {
-            $model = $this->menuRepository->getById($id);
-            $this->registry->register(self::REGISTRY_CODE, $model);
-            $result = $this->resultFactory->create(ResultFactory::TYPE_PAGE);
-            $result->setActiveMenu('Snowdog_Menu::menus');
-            $result->getConfig()->getTitle()->prepend(__('Edit Menu %1', $model->getTitle()));
-            return $result;
-        } catch (NoSuchEntityException $e) {
-            $result = $this->resultRedirectFactory->create();
-            $result->setPath('*/*/index');
-            return $result;
+        $menuId = (int) $this->getRequest()->getParam(self::ID);
+        $menu = $this->getCurrentMenu();
+
+        if ($menuId && !$menu->getMenuId()) {
+            $this->messageManager->addErrorMessage(__('This menu no longer exists.'));
+            /** @var Redirect $resultRedirect */
+            $resultRedirect = $this->resultRedirectFactory->create();
+
+            return $resultRedirect->setPath('*/*/index');
         }
-    }
 
-    protected function _isAllowed()
-    {
-        return $this->_authorization->isAllowed('Snowdog_Menu::menus');
+        $this->coreRegistry->register(self::REGISTRY_CODE, $menu);
+
+        /** @var Page $resultPage */
+        $resultPage = $this->resultPageFactory->create();
+        $this->initPage($resultPage)->addBreadcrumb(
+            $menuId ? __('Edit Menu %1', $menu->getTitle()) : __('New Menu'),
```

Comment:
This is repeated, maybe we can create a variable and store it and use it in both placeS?

Reply:
Extracted to a variable.